### PR TITLE
IC-1307 add end of service report entities and persistence tests

### PR DESCRIPTION
**Pull request by rgforsyth**
## What does this pull request do?

Adds the database changes required to add end of service report data.

## What is the intent behind these changes?

Allow end of service report data to be persisted in the database.

### CODE REVIEW

**Comment by rgforsyth** [comment]
There are database changes going in before this change, therefore the migration version number will change before merging.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/EndOfServiceReportOutcome.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import javax.persistence.Embeddable
+import javax.persistence.EnumType
+import javax.persistence.Enumerated
+import javax.persistence.FetchType
+import javax.persistence.ManyToOne
+import javax.validation.constraints.NotNull
+
+@Embeddable
+data class EndOfServiceReportOutcome(
+  @NotNull @ManyToOne(fetch = FetchType.LAZY) val desiredOutcome: DesiredOutcome,
```
Comment (mahalrmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/EndOfServiceReportOutcome.kt:12]: Perhaps this should be @Fetch(FetchMode.JOIN)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/EndOfServiceReportOutcome.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  @Enumerated(EnumType.STRING)
+  val achievementLevel: AchievementLevel,
+  val progressionComments: String? = null,
```
Comment (mahalrmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/EndOfServiceReportOutcome.kt:15]: Isn't AchievementLevel @NotNull

File: src/main/resources/db/migration/V1_32__end_of_service_report.sql [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    additional_task_comments text,
+
+    constraint fk_end_of_service_report_id foreign key (end_of_service_report_id) references end_of_service_report,
+    constraint fk_desired_outcome_id foreign key (desired_outcome_id) references desired_outcome
+);
+
+
+alter table referral
+    add column end_of_service_report_id uuid,
+    add constraint fk_end_of_service_report_id foreign key (end_of_service_report_id) references end_of_service_report;
```
Comment (mahalrmoj) [inline src/main/resources/db/migration/V1_32__end_of_service_report.sql:28]: Should there be a unique key on referral.end_of_service_report_id to each referral has a distinct EoSR

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/EndOfServiceReport.kt [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  var submittedAt: OffsetDateTime? = null,
+  @ManyToOne @Fetch(FetchMode.JOIN) var submittedBy: AuthUser? = null,
+
+  var furtherInformation: String? = null,
+
+  // Outcomes
+  @ElementCollection
+  @CollectionTable(name = "end_of_service_report_outcome", joinColumns = [JoinColumn(name = "end_of_service_report_id")])
+  @NotNull val outcomes: MutableList<EndOfServiceReportOutcome> = mutableListOf(),
+
```
Comment (mahalrmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/EndOfServiceReport.kt:33]: @tomsmyers was suggesting that the use of Sets may be better from a performance viewpoint

**Review by mahalrmoj — APPROVED** [review]
Ryan to discuss sets vs lists with Tom

**Pull request merged**